libopae++/include/opaec++/dma_buffer.h

```diff
@@ -0,0 +1,248 @@
+    {
+        if ((offset < len_) && (virt_ != nullptr)) {
+            return *reinterpret_cast<T *>(virt_ + offset);
+        }
+        // TODO log/throw error
+
+        return T();
+    }
+
+    /** Write a T-sized block of memory to the given location.
+     * @param[in] value The value to write.
+     * @param[in] offset The byte offset from the start of the buffer.
+     */
+    template <typename T>
+    void write(const T &value, size_t offset)
+    {
+        if ((offset < len_) && (virt_ != nullptr)) {
+            *reinterpret_cast<T *>(virt_ + offset) = value;
+        }
+        // TODO log/throw error
+    }
+
+    using microseconds_t = std::chrono::microseconds;
+    using time_point_t = std::chrono::high_resolution_clock::time_point;
+    using duration_t = std::chrono::duration<double>;
+
+    /** Poll for a specific value to appear at a given buffer location.
+     *
+     * Loops on a memory location without explicitly yielding the processor.
+     * This API should be used only for time-critical scenarios, eg waiting
+     * on a DMA address to be updated by hardware.
+     *
+     * @param[in] offset The byte offset from the start of the buffer.
+     * @param[in] timeout The maximum time to wait in microseconds.
+     * @param[in] mask A bit mask to apply to the value read from the buffer.
+     * @param[in] value The expected value after applying the mask.
+     * @retval true If the expected value was observed.
+     * @retval false If the timeout expired before seeing the expected value.
+     */
+    template <typename T>
+    bool poll(size_t offset, microseconds_t timeout, T mask, T value) const
+    {
+        time_point_t start = std::chrono::high_resolution_clock::now();
+        microseconds_t elapsed;
+
+        do
+        {
+            if ((read<T>(offset) & mask) == value)
+                return true;
+
+            elapsed = std::chrono::duration_cast<microseconds_t>(
+                        std::chrono::high_resolution_clock::now() - start);
+
+        }while(elapsed < timeout);
+
+        return false;
+    }
+
+    /** Wait for a specific value to appear at a given buffer location.
+     *
+     * Loops on a memory location, yielding the processor after each iteration.
+     * This API should not be used for time-critical scenarios. See poll instead.
+     *
+     * @param[in] offset The byte offset from the start of the buffer.
+     * @param[in] each The number of microseconds to sleep each loop iteration.
+     * @param[in] timeout The maximum time to wait in microseconds.
+     * @param[in] mask A bit mask to apply to the value read from the buffer.
+     * @param[in] value The expected value after applying the mask.
+     * @retval true If the expected value was observed.
+     * @retval false If the timeout expired before seeing the expected value.
+     */
+    template <typename T>
+    bool wait(size_t offset, microseconds_t each, microseconds_t timeout, T mask, T value) const
```

Comment:
I'm a little hesitant in including `poll` and `wait` in this API. What do you think? Should we keep this API simpler than that?